src/dbnode/storage/namespace/schema.go

```diff
@@ -0,0 +1,226 @@
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package namespace
+
+import (
+	"bytes"
+	"compress/gzip"
+	"errors"
+	"io/ioutil"
+	"github.com/golang/protobuf/proto"
+	dpb "github.com/golang/protobuf/protoc-gen-go/descriptor"
+	"github.com/jhump/protoreflect/desc"
+	nsproto "github.com/m3db/m3/src/dbnode/generated/proto/namespace"
+	xerrors "github.com/m3db/m3x/errors"
+	"strings"
+)
+
+var (
+	errInvalidSchema        = errors.New("invalid schema definition")
+	errSchemaNotFound       = errors.New("schema is not found")
+	errSchemaRegistryEmpty  = errors.New("schema registry is empty")
+	errInvalidSchemaOptions = errors.New("invalid schema options")
+)
+
+type schemaDescr struct {
+	deployId string
+	md       *desc.MessageDescriptor
+}
+
+func (s *schemaDescr) DeployId() string {
+	return s.deployId
+}
+
+func (s *schemaDescr) Equal(o SchemaDescr) bool {
+	if s == nil && o == nil {
+		return true
+	}
+	if s != nil && o == nil || s == nil && o != nil {
+		return false
+	}
+	if _, ok := o.(*schemaDescr); !ok {
+		return false
+	}
+	return s.DeployId() == o.DeployId()
+}
+
+func (s *schemaDescr) Get() *desc.MessageDescriptor {
+	return s.md
+}
+
+func (s *schemaDescr) String() string {
+	if s.md == nil {
+		return ""
+	}
+	return s.md.String()
+}
+
+type schemaRegistry struct {
+	options  *nsproto.SchemaOptions
+	latestId string
+	// a map of schema version to schema descriptor.
+	versions map[string]*schemaDescr
+}
+
+func (sr *schemaRegistry) Equal(o SchemaRegistry) bool {
+	var osr *schemaRegistry
+	var ok bool
+	if sr == nil && o == nil {
+		return true
+	}
+	if sr != nil && o == nil || sr == nil && o != nil {
+		return false
+	}
+
+	if osr, ok = o.(*schemaRegistry); !ok {
+		return false
+	}
+	// compare latest version
+	if sr.latestId != osr.latestId {
+		return false
+	}
+
+	// compare version map
+	if len(sr.versions) != len(osr.versions) {
+		return false
+	}
+	for v, sd := range sr.versions {
+		osd, ok := osr.versions[v]
+		if !ok {
+			return false
+		}
+		if !sd.Equal(osd) {
+			return false
+		}
+	}
+
+	return true
+}
+
+func (sr *schemaRegistry) Get(id string) (SchemaDescr, error) {
+	sd, ok := sr.versions[id]
+	if !ok {
+		return nil, errSchemaNotFound
+	}
+	return sd, nil
+}
+
+func (sr *schemaRegistry) GetLatest() (SchemaDescr, error) {
+	return sr.Get(sr.latestId)
+}
+
+// toSchemaOptions returns the corresponding SchemaOptions proto for the provided SchemaRegistry
+func toSchemaOptions(sr SchemaRegistry) *nsproto.SchemaOptions {
+	if sr == nil {
+		return nil
+	}
+	_, ok := sr.(*schemaRegistry)
+	if !ok {
+		return nil
+	}
+	return sr.(*schemaRegistry).options
+}
+
+func emptySchemaRegistry() SchemaRegistry {
+	return &schemaRegistry{options: nil, versions: make(map[string]*schemaDescr)}
+}
+
+// LoadSchemaRegistry loads schema registry from SchemaOptions proto.
```

Comment:
I think convention in go is usually to begin a constructor with `New`

Reply:
I prefer to reserve NewXXX for simpler constructor. This has logic to "load" from file descripitors, hence "Load".